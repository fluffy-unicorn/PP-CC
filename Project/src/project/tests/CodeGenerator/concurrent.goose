lock l;
shared int amount = 0;

thread add() {
	acquire l;
	print 'S';
	amount = amount + 100;
	print amount;
	print 'E';
	release l;
}

thread sub() {
	bool written = false;
	while(!written) {
		acquire l;
		print 'S';
		if(amount - 50 < 0) {
			print amount;
		}
		else {
			amount = amount - 50;
			print amount;
			written = true;
		}
		print 'E';	
		release l;
	}
}

function void main() {
	thread a = fork sub();
	#longComputation();
	thread b = fork add();
	join a;
	join b;
	
}

# function void longComputation() {
# 	int i = 0;
# 	while(i < 10)
#		i = i + 1;
# }




