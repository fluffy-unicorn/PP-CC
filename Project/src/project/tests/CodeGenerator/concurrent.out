Sprockell 1 says S
Sprockell 1 says 0
Sprockell 1 says E
Sprockell 2 says S
Sprockell 2 says 100
Sprockell 2 says E
Sprockell 1 says S
Sprockell 1 says 50
Sprockell 1 says E
